File: webviewexample/src/main/java/tech/thdev/webviewexample/view/main/MainFragment.java
package tech.thdev.webviewexample.view.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.widget.ContentLoadingProgressBar;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.webkit.WebSettings;
import android.widget.EditText;
import android.widget.TextView;

import org.jetbrains.annotations.Nullable;

import butterknife.BindView;
import tech.thdev.kotlin_example_01.base.view.BaseFragment;
import tech.thdev.webviewexample.R;
import tech.thdev.webviewexample.view.main.presenter.MainContract;
import tech.thdev.webviewexample.webkit.CustomWebChromeClient;
import tech.thdev.webviewexample.webkit.CustomWebView;
import tech.thdev.webviewexample.webkit.CustomWebViewClient;
import tech.thdev.webviewjavascriptinterface.util.ActivityUtilKt;

/**
 * Created by tae-hwan on 8/14/16.
 */

public class MainFragment extends BaseFragment<MainContract.Presenter> implements MainContract.View, TextView.OnEditorActionListener, View.OnKeyListener {

    @BindView(R.id.web_view)
    CustomWebView webView;

    private EditText etUrlView;
    private ContentLoadingProgressBar loadingProgressBar;

    private FloatingActionButton floatingActionButton;

    public static MainFragment getInstance() {
        return new MainFragment();
    }

    @Override
    public int getLayout() {
        return R.layout.fragment_main;
    }

    @Override
    public void onViewCreated(@Nullable View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initView();
    }

    private void initView() {
        etUrlView = (EditText) getActivity().findViewById(R.id.et_url_input);
        etUrlView.setOnEditorActionListener(this);
        floatingActionButton = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        floatingActionButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                webView.reload();
            }
        });

        loadingProgressBar = (ContentLoadingProgressBar) getActivity().findViewById(R.id.web_view_load_progress_bar);

//        webView.setOnKeyListener(this);

        // Add WebView ScrollChangeListener M
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
//            webView.setOnScrollChangeListener(new View.OnScrollChangeListener() {
//                @Override
//                public void onScrollChange(View view, int i, int i1, int i2, int i3) {
//
//                }
//            });
//        }

        webView.setOnWebViewListener(getPresenter());

        webView.setWebChromeClient(new CustomWebChromeClient(getActivity(), getPresenter()));
        webView.setWebViewClient(new CustomWebViewClient(getPresenter()));
        webView.defaultInit(WebSettings.LOAD_DEFAULT);
        if (getPresenter() != null) {
            getPresenter().defaultLoadUrl();
        }
    }

    @Override
    public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {
        switch (i) {
            case EditorInfo.IME_ACTION_GO:
                validationUrl();
                return true;

            default:
                break;
        }
        return false;
    }

    @Override
    public boolean onKey(View view, int i, KeyEvent keyEvent) {
        if (keyEvent.getAction() == KeyEvent.ACTION_DOWN) {
            switch (i) {
                case KeyEvent.KEYCODE_BACK:
                    if (webView.canGoBack()) {
                        webView.goBack();
                        return true;
                    }
                    break;
            }
        }
        return false;
    }

    /**
     * WebView Can go back.
     */
    public boolean canGoBack() {
        return webView.canGoBack();
    }

    /**
     * WebView go back.
     */
    public void goBack() {
        webView.goBack();
    }

    private void validationUrl() {
        if (getPresenter() != null) {
            getPresenter().validationUrl(etUrlView.getText().toString());
        }
    }

    @Override
    public void loadUrl(String url) {
        webView.loadUrl(url);
        ActivityUtilKt.hideKeyboard(getContext(), etUrlView);
    }

    @Override
    public void setUrl(String url) {
        etUrlView.setText(url);
    }

    @Override
    public void webViewScrollChanged(boolean isTop) {
        if (isTop) {
            floatingActionButton.show();
        } else {
            floatingActionButton.hide();
        }
    }

    @Override
    public void webViewProgressChanged(int newProgress) {
        loadingProgressBar.setProgress(newProgress);

        if (newProgress >= 100) {
            loadingProgressBar.setVisibility(View.GONE);
        } else if (loadingProgressBar.getVisibility() != View.VISIBLE) {
            loadingProgressBar.setVisibility(View.VISIBLE);
        }
    }

    public void urlShare() {
        Intent sendIntent = new Intent(Intent.ACTION_SEND);
        sendIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
        sendIntent.putExtra(Intent.EXTRA_SUBJECT, webView.getTitle());
        sendIntent.putExtra(Intent.EXTRA_TEXT, webView.getUrl());
        sendIntent.setType("text/plain");
        startActivity(Intent.createChooser(sendIntent, "Share link!"));
    }
}

File: app/src/main/java/tech/thdev/app/ui/main/MainFragment.kt
package tech.thdev.app.ui.main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.webkit.WebSettings
import androidx.appcompat.widget.AppCompatEditText
import androidx.core.widget.ContentLoadingProgressBar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import tech.thdev.app.R
import tech.thdev.app.base.view.BaseFragment
import tech.thdev.app.databinding.FragmentMainBinding
import tech.thdev.app.ui.main.presenter.MainContract
import tech.thdev.app.util.hideKeyboard
import tech.thdev.app.webkit.CustomWebChromeClient
import tech.thdev.app.webkit.CustomWebViewClient

/**
 * Created by tae-hwan on 8/14/16.
 */
class MainFragment : BaseFragment<MainContract.Presenter>(), MainContract.View {

    companion object {
        fun newInstance(): MainFragment =
            MainFragment()
    }

    private lateinit var binding: FragmentMainBinding

    private val etUrlView: AppCompatEditText by lazy {
        requireActivity().findViewById<AppCompatEditText>(R.id.et_url_input)
    }

    private val floatingActionButton: FloatingActionButton by lazy {
        requireActivity().findViewById<FloatingActionButton>(R.id.fab)
    }

    private val loadingProgressBar: ContentLoadingProgressBar by lazy {
        requireActivity().findViewById<ContentLoadingProgressBar>(R.id.web_view_load_progress_bar)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return FragmentMainBinding.inflate(inflater, container, false).also {
            binding = it
        }.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
    }

    private fun initView() {
        etUrlView.setOnEditorActionListener { _, actionId, _ ->
            when (actionId) {
                EditorInfo.IME_ACTION_GO -> {
                    validationUrl()
                    true
                }
                else -> false
            }
        }
        floatingActionButton.setOnClickListener {
            binding.webView.reload()
        }

        binding.webView.run {
            setOnKeyListener { _, keyCode, event ->
                if (event.action == KeyEvent.ACTION_DOWN) {
                    when (keyCode) {
                        KeyEvent.KEYCODE_BACK -> if (canGoBack()) {
                            goBack()
                            return@setOnKeyListener true
                        }
                    }
                }
                false
            }

            // Add WebView ScrollChangeListener M
            binding.webView.setOnScrollChangeListener { _, scrollX, scrollY, oldScrollX, oldScrollY ->
                Log.d(
                    "TAG",
                    "ScrollX $scrollX, ScrollY $scrollY oldScrollX $oldScrollX oldScrollY $oldScrollY"
                )
            }

            setOnWebViewListener(presenter)
            webChromeClient = CustomWebChromeClient(requireContext(), presenter)
            webViewClient = CustomWebViewClient(presenter)
            defaultInit(WebSettings.LOAD_DEFAULT)
        }

        presenter.defaultLoadUrl()
    }

    /**
     * WebView Can go back.
     */
    fun canGoBack(): Boolean {
        return binding.webView.canGoBack()
    }

    /**
     * WebView go back.
     */
    fun goBack() {
        binding.webView.goBack()
    }

    private fun validationUrl() {
        presenter.validationUrl(etUrlView.text.toString())
    }

    override fun loadUrl(url: String) {
        binding.webView.loadUrl(url)
        etUrlView.hideKeyboard()
    }

    override fun setUrl(url: String?) {
        etUrlView.setText(url)
    }

    override fun webViewScrollChanged(isTop: Boolean) {
        if (isTop) {
            floatingActionButton.show()
        } else {
            floatingActionButton.hide()
        }
    }

    override fun webViewProgressChanged(newProgress: Int) {
        loadingProgressBar.progress = newProgress
        if (newProgress >= 100) {
            loadingProgressBar.visibility = View.GONE
        } else if (loadingProgressBar.visibility != View.VISIBLE) {
            loadingProgressBar.visibility = View.VISIBLE
        }
    }

    fun urlShare() {
        val sendIntent = Intent(Intent.ACTION_SEND)
        sendIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
        sendIntent.putExtra(Intent.EXTRA_SUBJECT, binding.webView.title)
        sendIntent.putExtra(Intent.EXTRA_TEXT, binding.webView.url)
        sendIntent.type = "text/plain"
        startActivity(Intent.createChooser(sendIntent, "Share link!"))
    }
}